File: apps/macos/Sources/CornerTasks/Sync/SyncEngine.swift
import CryptoKit
import Foundation

/// End-to-end-encrypted sync engine for the macOS app. Lifecycle:
///
/// - Construct only when `Prefs.cloudSyncEnabled == true` and a valid `backendURL` and
///   `Identity` are available. While disabled, no instance exists — therefore no timers,
///   no enqueuer, and no network calls.
/// - `start()` installs the `TaskStore.eventBuilder` that writes encrypted queue rows on
///   every local mutation, kicks off an immediate `flushPushes()`, and schedules the
///   10-minute push and 1-minute pull timers.
/// - `stop()` cancels both timers and clears the enqueuer. Local state is untouched.
///
/// The 60-day archive cutoff is enforced at flush time: any pending row whose task has
/// `completed_at < now - 60 d` is marked sent without being POSTed.
final class SyncEngine {
    static let archiveCutoff: TimeInterval = 60 * 24 * 60 * 60  // 60 days

    private let store: TaskStore
    private let transport: SyncTransport
    private let auth: AuthSession
    private let identity: Identity
    private let encryptionKey: SymmetricKey
    private let deviceId: String
    private let now: () -> Date
    private let lastSyncedAtStorage: LastSyncedAtStorage
    /// Reads the user-configured sync cadence, in seconds. Defaults to `Prefs.syncIntervalSeconds`.
    private let intervalSeconds: () -> TimeInterval
    /// Set + cleared by the engine. Persisted in `Prefs.pendingFullResync` for callers
    /// that want to trigger a resync on the next start (typically after the user
    /// (re-)enables cloud sync).
    private let pendingFullResync: () -> Bool
    private let clearPendingFullResync: () -> Void

    private var pushTimer: Timer?
    private var pullTimer: Timer?

    init(store: TaskStore,
         transport: SyncTransport,
         identity: Identity,
         encryptionKey: SymmetricKey,
         deviceId: String,
         lastSyncedAt: LastSyncedAtStorage = UserDefaultsLastSyncedAt(),
         intervalSeconds: @escaping () -> TimeInterval = { TimeInterval(Prefs.syncIntervalSeconds) },
         pendingFullResync: @escaping () -> Bool = { Prefs.pendingFullResync },
         clearPendingFullResync: @escaping () -> Void = { Prefs.pendingFullResync = false },
         now: @escaping () -> Date = Date.init) {
        self.store = store
        self.transport = transport
        self.identity = identity
        self.encryptionKey = encryptionKey
        self.deviceId = deviceId
        self.now = now
        self.lastSyncedAtStorage = lastSyncedAt
        self.intervalSeconds = intervalSeconds
        self.pendingFullResync = pendingFullResync
        self.clearPendingFullResync = clearPendingFullResync
        self.auth = AuthSession(identity: identity, transport: transport, now: now)
    }

    /// Test hook.
    var authSession: AuthSession { auth }

    func start() {
        installEnqueuer()
        let interval = max(TimeInterval(Prefs.syncIntervalMinSeconds), intervalSeconds())
        let needsResync = pendingFullResync()
        Task {
            if needsResync {
                await self.performFullResync()
                self.clearPendingFullResync()
            } else {
                await self.flushPushes()
            }
        }
        let push = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task { await self?.flushPushes() }
        }
        let pull = Timer.scheduledTimer(withTimeInterval: interval, repeats: true) { [weak self] _ in
            Task { await self?.pullSince() }
        }
        pushTimer = push
        pullTimer = pull
    }

    func stop() {
        pushTimer?.invalidate(); pushTimer = nil
        pullTimer?.invalidate(); pullTimer = nil
        store.eventBuilder = nil
    }

    /// Sends every pending queue row that is not past the archive cutoff. Marks rows
    /// accepted or stale-rejected as sent. Auth failures cause one re-auth + retry per
    /// flush; persistent auth failure leaves rows in the queue for the next tick.
    func flushPushes() async {
        let rows = store.pendingQueueRows()
        guard !rows.isEmpty else { return }

        var toSend: [SyncEvent] = []
        var skippedSent: [String] = []
        let cutoff = now().addingTimeInterval(-Self.archiveCutoff)

        for row in rows {
            if row.op == "upsert",
               let completedAt = store.taskCompletedAt(taskId: row.taskId),
               completedAt < cutoff {
                skippedSent.append(row.eventId)
                continue
            }
            guard let event = try? JSONDecoder().decode(SyncEvent.self, from: row.payloadJSON) else {
                continue
            }
            toSend.append(event)
        }

        if !skippedSent.isEmpty {
            store.markQueueRowsSent(eventIds: skippedSent, at: now())
        }
        guard !toSend.isEmpty else { return }

        do {
            let response = try await pushWithAuth(events: toSend)
            let toMark = response.accepted + response.rejected.map(\.eventId)
            if !toMark.isEmpty {
                store.markQueueRowsSent(eventIds: toMark, at: now())
            }
        } catch {
            // Leave rows pending; retry next tick.
        }
    }

    /// One-shot reset. Discards every row in the local outbound queue, pulls the
    /// full state from the backend (since-epoch), then re-enqueues an upsert for
    /// every local task the server did not return — and pushes the lot. After this
    /// runs, the cloud holds an up-to-date copy of every live local task. Archived
    /// tasks completed more than `archiveCutoff` seconds ago are intentionally
    /// left out; they are dropped from both ends by the retention policy.
    func performFullResync() async {
        store.clearSyncQueue()
        lastSyncedAtStorage.value = ISO8601.format(Date(timeIntervalSince1970: 0))
        let pulled = await pullForResync()
        let cutoff = now().addingTimeInterval(-Self.archiveCutoff)
        for snapshot in store.liveSnapshots() {
            if pulled.contains(snapshot.taskId.uuidString) { continue }
            if let completed = snapshot.completedAt, completed < cutoff { continue }
            store.enqueueSnapshot(snapshot)
        }
        await flushPushes()
    }

    private func pullForResync() async -> Set<String> {
        let since = lastSyncedAtStorage.value ?? defaultSince()
        do {
            let response = try await pullWithAuth(since: since)
            var pulled: Set<String> = []
            for event in response.events {
                applyRemote(event)
                pulled.insert(event.taskId)
            }
            lastSyncedAtStorage.value = response.serverTime
            return pulled
        } catch {
            return []
        }
    }

    /// Fetches events with `updatedAt >= lastSyncedAt`, applies them last-writer-wins,
    /// then advances `lastSyncedAt` to the server's clock from the response.
    func pullSince() async {
        let since = lastSyncedAtStorage.value ?? defaultSince()
        do {
            let response = try await pullWithAuth(since: since)
            for event in response.events {
                applyRemote(event)
            }
            lastSyncedAtStorage.value = response.serverTime
        } catch {
            // Network or auth failure — try again next tick.
        }
    }

    // MARK: - Bearer-aware push/pull

    private func pushWithAuth(events: [SyncEvent]) async throws -> PushResponse {
        let token = try await auth.bearer()
        do {
            return try await transport.push(accountDid: identity.accountDid, events: events, bearer: token)
        } catch SyncTransportError.tokenExpired, SyncTransportError.badToken {
            await auth.invalidate()
            let fresh = try await auth.bearer()
            return try await transport.push(accountDid: identity.accountDid, events: events, bearer: fresh)
        }
    }

    private func pullWithAuth(since: String) async throws -> PullResponse {
        let token = try await auth.bearer()
        do {
            return try await transport.pull(accountDid: identity.accountDid, since: since, bearer: token)
        } catch SyncTransportError.tokenExpired, SyncTransportError.badToken {
            await auth.invalidate()
            let fresh = try await auth.bearer()
            return try await transport.pull(accountDid: identity.accountDid, since: since, bearer: fresh)
        }
    }

    // MARK: - Enqueue (local mutations → encrypted queue rows)

    /// Internal so tests can install the enqueuer without scheduling real timers.
    func installEnqueuer() {
        let key = encryptionKey
        let did = identity.accountDid
        let device = deviceId
        store.eventBuilder = { snapshot in
            let eventId = UUID().uuidString.lowercased()
            let plaintext: EventPlaintext?
            if snapshot.op == "delete" {
                plaintext = nil
            } else {
                plaintext = EventPlaintext(
                    title: snapshot.title,
                    createdAt: snapshot.createdAt,
                    completedAt: snapshot.completedAt,
                    dueDate: snapshot.dueDate,
                    order: snapshot.order
                )
            }
            do {
                let event = try SyncEventCodec.makeEvent(
                    op: snapshot.op,
                    accountDid: did,
                    deviceId: device,
                    eventId: eventId,
                    taskId: snapshot.taskId.uuidString,
                    updatedAt: snapshot.updatedAt,
                    plaintext: plaintext,
                    encryptionKey: key
                )
                let json = try JSONEncoder().encode(event)
                return BuiltSyncEvent(eventId: eventId, payloadJSON: json)
            } catch {
                return nil
            }
        }
    }

    // MARK: - Apply pulled events

    private func applyRemote(_ event: SyncEvent) {
        guard event.accountDid == identity.accountDid else { return }
        guard let taskId = UUID(uuidString: event.taskId) else { return }
        guard let updatedAt = ISO8601.parse(event.updatedAt) else { return }

        // Defense in depth: events for archived tasks older than 60 days are ignored.
        let cutoff = now().addingTimeInterval(-Self.archiveCutoff)
        if updatedAt < cutoff && event.op == "upsert" { return }

        do {
            let plaintext = try SyncEventCodec.openEvent(event, encryptionKey: encryptionKey)
            switch event.op {
            case "upsert":
                guard let pt = plaintext else { return }
                store.applyRemoteUpsert(
                    taskId: taskId,
                    title: pt.title,
                    createdAt: pt.createdAt,
                    completedAt: pt.completedAt,
                    dueDate: pt.dueDate,
                    order: pt.order,
                    updatedAt: updatedAt,
                    eventId: event.eventId
                )
            case "delete":
                store.applyRemoteDelete(taskId: taskId, updatedAt: updatedAt, eventId: event.eventId)
            default:
                return
            }
        } catch {
            return
        }
    }

    private func defaultSince() -> String {
        ISO8601.format(Date(timeIntervalSince1970: 0))
    }
}

// MARK: - lastSyncedAt persistence

protocol LastSyncedAtStorage: AnyObject {
    var value: String? { get set }
}

final class UserDefaultsLastSyncedAt: LastSyncedAtStorage {
    private let key = "cornertasks.sync.lastSyncedAt"
    var value: String? {
        get { UserDefaults.standard.string(forKey: key) }
        set {
            if let v = newValue { UserDefaults.standard.set(v, forKey: key) }
            else { UserDefaults.standard.removeObject(forKey: key) }
        }
    }
}

final class InMemoryLastSyncedAt: LastSyncedAtStorage {
    var value: String?
    init(_ initial: String? = nil) { self.value = initial }
}

File: apps/web/src/sync/SyncEngine.ts
// End-to-end-encrypted sync engine for the web app. Mirrors apps/macos/Sources/CornerTasks/Sync/SyncEngine.swift.
//
// Lifecycle:
// - Constructed only when cloud sync is enabled, a backendURL is set, and an Identity exists.
//   While disabled, no instance exists → no eventBuilder installed → no syncQueue rows → no
//   network calls.
// - `start()` installs a `TaskStore.eventBuilder` that writes encrypted queue rows on every
//   local mutation, immediately runs `flushPushes()` and `pullSince()`, and schedules the
//   10-minute push and 1-minute pull intervals. While `document.hidden`, the periodic timers
//   are paused; `visibilitychange → visible` resumes them and runs both immediately.
// - `stop()` clears intervals, removes the eventBuilder, releases the visibility listener.

import type { Identity } from '../crypto/identity';
import type { TaskStore, TaskMutationSnapshot, BuiltSyncEvent, EventBuilder } from '../storage/TaskStore';
import { isoParse, makeEvent, openEvent, type SyncEvent } from './syncEvent';
import { AuthSession } from './authSession';
import type { SyncTransport, SyncTransportError } from './syncTransport';
import { isSyncTransportError } from './syncTransport';

export interface LastSyncedAtStorage {
  get(): string | null;
  set(v: string | null): void;
}

export interface VisibilityHost {
  isHidden(): boolean;
  addListener(fn: () => void): () => void;
}

export const ARCHIVE_CUTOFF_MS = 60 * 24 * 60 * 60 * 1000;

/** Interim fix for the pull-cursor race: events whose client-set `updatedAt`
 *  is older than the server's `serverTime` at our last pull but which arrive
 *  at the server *after* that pull would otherwise be lost forever. We rewind
 *  `lastSyncedAt` by `CURSOR_LOOKBACK_MS` on every advance so those events
 *  come back on the next round. `applyRemote` is idempotent under
 *  last-writer-wins, so the redelivery is harmless. Replace with a
 *  server-assigned monotonic cursor in a follow-up iteration; see README
 *  "Known limitations". */
export const CURSOR_LOOKBACK_MS = 5 * 60 * 1000;

const cursorAfter = (serverTime: string): string => {
  const t = Date.parse(serverTime);
  if (Number.isNaN(t)) return serverTime;
  return new Date(Math.max(0, t - CURSOR_LOOKBACK_MS)).toISOString();
};
/** Test/legacy default. Real cadence is now `Prefs.syncIntervalSeconds`. */
export const DEFAULT_SYNC_INTERVAL_MS = 60_000;

const LS_LAST_SYNCED = 'cornertasks.sync.lastSyncedAt';
const LS_DEVICE_ID = 'cornertasks.sync.deviceId';
const EPOCH_ZERO_ISO = new Date(0).toISOString();

export const localStorageLastSyncedAt = (): LastSyncedAtStorage => ({
  get: () => localStorage.getItem(LS_LAST_SYNCED),
  set: (v) => {
    if (v === null) localStorage.removeItem(LS_LAST_SYNCED);
    else localStorage.setItem(LS_LAST_SYNCED, v);
  },
});

export const inMemoryLastSyncedAt = (initial?: string): LastSyncedAtStorage => {
  let v: string | null = initial ?? null;
  return { get: () => v, set: (x) => { v = x; } };
};

export const ensureDeviceId = (): string => {
  const existing = localStorage.getItem(LS_DEVICE_ID);
  if (existing) return existing;
  const id =
    typeof crypto !== 'undefined' && 'randomUUID' in crypto
      ? crypto.randomUUID()
      : 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
          const r = (Math.random() * 16) | 0;
          return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
        });
  localStorage.setItem(LS_DEVICE_ID, id);
  return id;
};

export const documentVisibilityHost = (): VisibilityHost => ({
  isHidden: () => typeof document !== 'undefined' && document.hidden,
  addListener: (fn) => {
    if (typeof document === 'undefined') return () => {};
    document.addEventListener('visibilitychange', fn);
    return () => document.removeEventListener('visibilitychange', fn);
  },
});

export const noopVisibilityHost = (): VisibilityHost => ({
  isHidden: () => false,
  addListener: () => () => {},
});

export const CLOUD_SYNC_CHANGED_EVENT = 'cornertasks:cloud-sync-changed';
/** Lightweight signal: only the timer cadence changed. The current engine
 *  reschedules its timers and keeps its bearer + AuthSession. Distinct from
 *  `CLOUD_SYNC_CHANGED_EVENT`, which is a full restart trigger. */
export const CLOUD_SYNC_INTERVAL_CHANGED_EVENT = 'cornertasks:cloud-sync-interval-changed';

export const fireCloudSyncChanged = (enabled: boolean): void => {
  if (typeof window === 'undefined') return;
  window.dispatchEvent(new CustomEvent(CLOUD_SYNC_CHANGED_EVENT, { detail: { enabled } }));
};

export const fireCloudSyncIntervalChanged = (): void => {
  if (typeof window === 'undefined') return;
  window.dispatchEvent(new CustomEvent(CLOUD_SYNC_INTERVAL_CHANGED_EVENT));
};

export interface SyncEngineOptions {
  store: TaskStore;
  transport: SyncTransport;
  identity: Identity;
  encryptionKey: Uint8Array;
  deviceId: string;
  lastSyncedAt?: LastSyncedAtStorage;
  visibility?: VisibilityHost;
  now?: () => Date;
  /** User-configured cadence in ms; same value drives push and pull timers. */
  intervalMs?: () => number;
  /** Returns true once the user has just (re-)enabled cloud sync. The engine
   *  consumes the flag exactly once via `clearPendingFullResync`. */
  pendingFullResync?: () => boolean;
  clearPendingFullResync?: () => void;
  /** Test seam: replaces window.setInterval / clearInterval. */
  setInterval?: (fn: () => void, ms: number) => unknown;
  clearInterval?: (handle: unknown) => void;
}

export class SyncEngine {
  private readonly store: TaskStore;
  private readonly transport: SyncTransport;
  private readonly identity: Identity;
  private readonly encryptionKey: Uint8Array;
  private readonly deviceId: string;
  private readonly lastSyncedAt: LastSyncedAtStorage;
  private readonly visibility: VisibilityHost;
  private readonly now: () => Date;
  private readonly setIntervalImpl: (fn: () => void, ms: number) => unknown;
  private readonly clearIntervalImpl: (h: unknown) => void;
  private readonly intervalMs: () => number;
  private readonly pendingFullResyncFn: () => boolean;
  private readonly clearPendingFullResyncFn: () => void;
  readonly authSession: AuthSession;

  private pushTimer: unknown = null;
  private pullTimer: unknown = null;
  private removeVisibilityListener: (() => void) | null = null;

  constructor(opts: SyncEngineOptions) {
    this.store = opts.store;
    this.transport = opts.transport;
    this.identity = opts.identity;
    this.encryptionKey = opts.encryptionKey;
    this.deviceId = opts.deviceId;
    this.lastSyncedAt = opts.lastSyncedAt ?? localStorageLastSyncedAt();
    this.visibility = opts.visibility ?? documentVisibilityHost();
    this.now = opts.now ?? (() => new Date());
    this.setIntervalImpl =
      opts.setInterval ??
      ((fn, ms) => (typeof window !== 'undefined' ? window.setInterval(fn, ms) : setInterval(fn, ms)));
    this.clearIntervalImpl =
      opts.clearInterval ??
      ((h) => {
        if (typeof window !== 'undefined') window.clearInterval(h as number);
        else clearInterval(h as ReturnType<typeof setInterval>);
      });
    this.intervalMs = opts.intervalMs ?? (() => DEFAULT_SYNC_INTERVAL_MS);
    this.pendingFullResyncFn = opts.pendingFullResync ?? (() => false);
    this.clearPendingFullResyncFn = opts.clearPendingFullResync ?? (() => {});
    this.authSession = new AuthSession(this.identity, this.transport, { now: this.now });
  }

  start(): void {
    this.installEventBuilder();
    if (this.pendingFullResyncFn()) {
      void this.performFullResync().then(() => this.clearPendingFullResyncFn());
    } else {
      void this.flushPushes();
      void this.pullSince();
    }
    if (!this.visibility.isHidden()) this.scheduleTimers();
    this.removeVisibilityListener = this.visibility.addListener(() => this.onVisibilityChange());
  }

  stop(): void {
    this.cancelTimers();
    this.store.eventBuilder = null;
    if (this.removeVisibilityListener) {
      this.removeVisibilityListener();
      this.removeVisibilityListener = null;
    }
  }

  /** Internal but exposed so tests can install the builder without scheduling intervals. */
  installEventBuilder(): void {
    const builder: EventBuilder = async (snapshot: TaskMutationSnapshot): Promise<BuiltSyncEvent | null> => {
      const eventId = newEventId();
      const plaintext =
        snapshot.op === 'delete'
          ? null
          : {
              title: snapshot.title,
              createdAt: snapshot.createdAt,
              completedAt: snapshot.completedAt,
              dueDate: snapshot.dueDate,
              order: snapshot.order,
            };
      try {
        const event = await makeEvent({
          op: snapshot.op,
          accountDid: this.identity.accountDid,
          deviceId: this.deviceId,
          eventId,
          // Canonical lowercase per docs/sync-protocol.md §3. crypto.randomUUID
          // already lowercases, but old web rows or imports could still hold an
          // uppercase id; lowercase here so wire events are always canonical.
          taskId: snapshot.taskId.toLowerCase(),
          updatedAt: snapshot.updatedAt,
          plaintext,
          encryptionKey: this.encryptionKey,
        });
        return { eventId, payloadJSON: JSON.stringify(event) };
      } catch {
        return null;
      }
    };
    this.store.eventBuilder = builder;
  }

  async flushPushes(): Promise<void> {
    const rows = await this.store.pendingQueueRows();
    if (rows.length === 0) return;

    const cutoff = this.now().getTime() - ARCHIVE_CUTOFF_MS;
    const toSend: SyncEvent[] = [];
    const skippedSent: string[] = [];

    for (const row of rows) {
      if (row.op === 'upsert') {
        const completedAt = await this.store.taskCompletedAt(row.taskId);
        if (completedAt && completedAt.getTime() < cutoff) {
          skippedSent.push(row.eventId);
          continue;
        }
      }
      try {
        toSend.push(JSON.parse(row.payloadJSON) as SyncEvent);
      } catch {
        // Corrupt row — skip silently; the next mutation will replace it.
      }
    }

    if (skippedSent.length > 0) await this.store.markQueueRowsSent(skippedSent);
    if (toSend.length === 0) return;

    try {
      const response = await this.pushWithAuth(toSend);
      const toMark = [...response.accepted, ...response.rejected.map((r) => r.eventId)];
      if (toMark.length > 0) await this.store.markQueueRowsSent(toMark);
    } catch (err) {
      console.warn('[CornerTasks sync] push failed; will retry next tick', err);
    }
  }

  /** One-shot reset. Discards every row in the local outbound queue, pulls the
   *  full state from the backend (since-epoch), then re-enqueues an upsert for
   *  every local task the server did not return — and pushes the lot. After this
   *  runs, the cloud holds an up-to-date copy of every live local task. Archived
   *  tasks completed more than `ARCHIVE_CUTOFF_MS` ago are intentionally left out
   *  — they are dropped from both ends by the retention policy. */
  async performFullResync(): Promise<void> {
    await this.store.clearSyncQueue();
    this.lastSyncedAt.set(EPOCH_ZERO_ISO);
    const pulled = await this.pullForResync();
    const cutoff = this.now().getTime() - ARCHIVE_CUTOFF_MS;
    for (const snapshot of await this.store.liveSnapshots()) {
      if (pulled.has(snapshot.taskId)) continue;
      if (snapshot.completedAt && snapshot.completedAt.getTime() < cutoff) continue;
      await this.store.enqueueSnapshot(snapshot);
    }
    await this.flushPushes();
  }

  private async pullForResync(): Promise<Set<string>> {
    const since = this.lastSyncedAt.get() ?? EPOCH_ZERO_ISO;
    try {
      const response = await this.pullWithAuth(since);
      const pulled = new Set<string>();
      for (const event of response.events) {
        await this.applyRemote(event);
        pulled.add(event.taskId);
      }
      this.lastSyncedAt.set(cursorAfter(response.serverTime));
      console.info('[CornerTasks sync] resync pull complete', { pulled: pulled.size });
      return pulled;
    } catch (err) {
      console.warn('[CornerTasks sync] resync pull failed; rebuilding queue from local only', err);
      return new Set();
    }
  }

  async pullSince(): Promise<void> {
    const since = this.lastSyncedAt.get() ?? EPOCH_ZERO_ISO;
    try {
      const response = await this.pullWithAuth(since);
      for (const event of response.events) {
        await this.applyRemote(event);
      }
      this.lastSyncedAt.set(cursorAfter(response.serverTime));
    } catch (err) {
      console.warn('[CornerTasks sync] pull failed; will retry next tick', err);
    }
  }

  // MARK: - Bearer-aware push/pull

  private async pushWithAuth(events: SyncEvent[]) {
    const token = await this.authSession.bearer();
    try {
      return await this.transport.push(this.identity.accountDid, events, token);
    } catch (e) {
      if (isAuthRetryable(e)) {
        this.authSession.invalidate();
        const fresh = await this.authSession.bearer();
        return await this.transport.push(this.identity.accountDid, events, fresh);
      }
      throw e;
    }
  }

  private async pullWithAuth(since: string) {
    const token = await this.authSession.bearer();
    try {
      return await this.transport.pull(this.identity.accountDid, since, token);
    } catch (e) {
      if (isAuthRetryable(e)) {
        this.authSession.invalidate();
        const fresh = await this.authSession.bearer();
        return await this.transport.pull(this.identity.accountDid, since, fresh);
      }
      throw e;
    }
  }

  // MARK: - Apply pulled events

  private async applyRemote(event: SyncEvent): Promise<void> {
    if (event.accountDid !== this.identity.accountDid) return;
    const updatedAt = isoParse(event.updatedAt);
    if (!updatedAt) return;

    // Defense-in-depth: ignore very old upserts.
    if (event.op === 'upsert' && updatedAt.getTime() < this.now().getTime() - ARCHIVE_CUTOFF_MS) return;

    let plaintext;
    try {
      plaintext = await openEvent(event, this.encryptionKey);
    } catch {
      return;
    }
    // Canonical lowercase per docs/sync-protocol.md §3. v0.2.0-pre macOS
    // builds emitted uppercase taskIds; lowercasing on receipt keeps
    // IndexedDB keys consistent across clients.
    const taskId = event.taskId.toLowerCase();
    if (event.op === 'upsert') {
      if (!plaintext) return;
      await this.store.applyRemoteUpsert({
        taskId,
        title: plaintext.title,
        createdAt: plaintext.createdAt,
        completedAt: plaintext.completedAt,
        dueDate: plaintext.dueDate,
        order: plaintext.order,
        updatedAt,
        eventId: event.eventId,
      });
    } else if (event.op === 'delete') {
      await this.store.applyRemoteDelete({ taskId, updatedAt, eventId: event.eventId });
    }
  }

  // MARK: - Tab visibility

  private onVisibilityChange(): void {
    if (this.visibility.isHidden()) {
      this.cancelTimers();
    } else {
      void this.flushPushes();
      void this.pullSince();
      this.scheduleTimers();
    }
  }

  /** Re-arm both timers at the current `intervalMs()` without an immediate
   *  push/pull and without rebuilding `AuthSession`. Use this when only the
   *  cadence changes — full engine recreate (which loses the cached bearer
   *  and re-auths) is overkill for a slider drag. */
  reschedule(): void {
    if (this.pushTimer === null && this.pullTimer === null) return;
    this.cancelTimers();
    if (!this.visibility.isHidden()) this.scheduleTimers();
  }

  private scheduleTimers(): void {
    const ms = Math.max(1000, this.intervalMs());
    if (this.pushTimer === null) {
      this.pushTimer = this.setIntervalImpl(() => { void this.flushPushes(); }, ms);
    }
    if (this.pullTimer === null) {
      this.pullTimer = this.setIntervalImpl(() => { void this.pullSince(); }, ms);
    }
  }

  private cancelTimers(): void {
    if (this.pushTimer !== null) { this.clearIntervalImpl(this.pushTimer); this.pushTimer = null; }
    if (this.pullTimer !== null) { this.clearIntervalImpl(this.pullTimer); this.pullTimer = null; }
  }
}

const newEventId = (): string => {
  if (typeof crypto !== 'undefined' && 'randomUUID' in crypto) return crypto.randomUUID();
  return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, (c) => {
    const r = (Math.random() * 16) | 0;
    return (c === 'x' ? r : (r & 0x3) | 0x8).toString(16);
  });
};

const isAuthRetryable = (e: unknown): boolean => {
  if (!isSyncTransportError(e)) return false;
  return e.kind === 'tokenExpired' || e.kind === 'badToken';
};

// Re-export so tests can inject without importing the type directly.
export type { SyncTransportError };
